src/history/hs.c

```diff
@@ -307,9 +307,22 @@ static int
 __hs_row_search(WT_CURSOR_BTREE *hs_cbt, WT_ITEM *srch_key, bool insert)
 {
     WT_DECL_RET;
+    bool leaf_found;
+
+    leaf_found = false;
+
+    /*
+     * Check whether the search key can be find in the provided leaf page, if exists. Otherwise
+     * perform a full search.
+     */
+    if (hs_cbt->ref != NULL)
+        WT_WITH_BTREE(CUR2S(hs_cbt), CUR2BT(hs_cbt),
+          ret = __wt_row_search(hs_cbt, srch_key, insert, hs_cbt->ref, false, &leaf_found));
+
+    if (!leaf_found)
+        WT_WITH_BTREE(CUR2S(hs_cbt), CUR2BT(hs_cbt),
+          ret = __wt_row_search(hs_cbt, srch_key, insert, NULL, false, NULL));
```

Comment:
I expected that you would need to call `cursor->reset`, or `__wt_page_release` in between the two searches if `leaf_found == false`.

Otherwise, I think the hazard pointer for the pinned page will be leaked.

Reply:
Thanks.  Fixed by calling the `cursor->reset()`, but still some tests are failed with cursor key out of order. Will check them and update.

Reply:
@michaelcahill Thanks for fixing the key out of order problem.